[3007.x] transports: fix and normalize PublishServer method signatures

Sxderp opened the pull request:
the arguments to publish, specifically io_loop, needs to be consistant across the different transports. This argument was not consistant and caused exceptions.

Sxderp committed:
zeromq: fix io_loop variable name in publish function

the arguments to publish, specifically io_loop, needs to be consistant
across the different transports. This argument was not consistant and
caused exceptions.

Sxderp committed:
update io_loop kwarg name for zeromq tests

Sxderp committed:
add test for PublishServer.publisher arguments

Sxderp commented:
I guess I never gave a full rundown on why this is needed.

The master creates a [publish server](https://github.com/saltstack/salt/blob/bd89384259c4fd577f60273b4fc6bb465b98e6f9/salt/master.py#L819-L822). The [factory method](https://github.com/saltstack/salt/blob/bd89384259c4fd577f60273b4fc6bb465b98e6f9/salt/channel/server.py#L968-L970) creates a `MasterPubServerChannel` with the chosen transport using [ipc_publish_server](https://github.com/saltstack/salt/blob/bd89384259c4fd577f60273b4fc6bb465b98e6f9/salt/transport/base.py#L213), we're going to come back to that function in a moment.

In the pre_fork we call `MasterPubServerChannel._publish_domain`, which calls the [publisher function on the transport](https://github.com/saltstack/salt/blob/bd89384259c4fd577f60273b4fc6bb465b98e6f9/salt/channel/server.py#L1065-L1069). If the arguments are not correct then the function call will fail.

Going back to [ipc_publish_server](https://github.com/saltstack/salt/blob/bd89384259c4fd577f60273b4fc6bb465b98e6f9/salt/transport/base.py#L213) (and `ipc_publish_client` right next to it). There is an arbitrary restriction on using TCP only for these functions. I'm willing to bet that's because /for some reason/ (cough) zeromq was not working and the solution was to just use TCP instead

After my patch I was able to remove the TCP restriction and everything worked.

I have also added a test for required arguments. Perhaps it may make more sense with abstract classes? I'm not really sure. I'm also not 100% on what folder / file should be modified for the tests.

Sxderp committed:
lint: follow code convention for new arg test in transport

Sxderp commented:
@dwoz I think this is fully ready. The CI is not running "core tests" so my test isn't actually executed. I'm not sure if there's something I need to do for that?

Sxderp committed:
use abstractmethods to test that publishserver is implemented correctly

Sxderp committed:
normalize transport PublishServer method signatures

Sxderp commented:
I have rebased onto 3007.x.

I have also extended the PR to include the other methods of PublishServer.

@dwoz
@dmurphy18 

A followup PR should probably be made for PublishClient..

Sxderp committed:
lint: follow code convention

Sxderp committed:
transport: remove write callback in tcp

Sxderp commented:
I think the write callback in transport.tcp is a holderver from IPCServer. I could not trace any code that actually used it. Local tests seem to agree.

Sxderp commented:
@dmurphy18 requesting Workflows get ran again.

Sxderp committed:
lint: declare publishing abstract methods as abstract

Sxderp committed:
test: pub channel remove callback and use new classes

TCP* classes are deprecated

Sxderp committed:
remove unnecessary argument from channel handler

Sxderp commented:
Small fixups to get tests to succeed.

Sxderp committed:
lint: add newline to test file

Sxderp commented:
Well, that's embarrassing. I missed a linter item. Quick fix incoming for that.

Sxderp committed:
lint: move imports from the same module onto one line

Sxderp commented:
Flaky tests? They work locally and were successful on the previous commit (before my lint fixup).

Sxderp commented:
Rebased onto 3007.x. No conflicts.

Sxderp commented:
As the upstream changes I keep rebasing and pushing to bring the branch up to date. I'm wondering if I should keep doing that or to leave the branch "as-is"?

twangboy commented:
I don't think you need to rebase for now... unless there is a conflict.

dwoz commented on salt/transport/base.py:
```diff
@@ -374,9 +376,23 @@ class DaemonizedPublishServer(PublishServer):
+        pull_path=None,
+        pull_path_perms=0o600,
+        pub_path_perms=0o600,
+        ssl=None,
```
ssl is not a valid option for all transports.

Sxderp replied:
Then it should be ignored by that transport. We already have [issues](https://github.com/saltstack/salt/issues/66577#issuecomment-2251245303) with trying to pass the 'ssl' config when a transport does not support it. Do you expect the [factory methods](https://github.com/saltstack/salt/blob/ca322bdd1d6cfc1f8d935120e387b6d3b8e09a83/salt/transport/base.py#L99-L119) to know which transports do and do not support SSL and not pass it along?

dwoz replied:
I'd argue that it's the whole point of using a factory method rather than just instantiating the class directly.

Sxderp replied:
Updated so the factory method checks for SSL.

Sxderp committed:
update transport factory to check for ssl support

dwoz commented:
@Sxderp Can you rebase this please?

Sxderp committed:
change support_ssl to a class method

it is easier to do abstractmethod with standard class methods rather
than properties

Sxderp committed:
all additiona parameters for __init__ for publishserver subclasses

Sxderp committed:
lint: cleanup errors in transport methods

Sxderp committed:
zeromq: use consistent name for io_loop variable

Sxderp commented:
Fixed lint and variable name (must have got missed in a rebase or something..).

Sxderp commented:
That was weird. Not sure how that happened. Did a fresh rebase and it's all fixed.

Sxderp committed:
lint: cleanup

Sxderp committed:
make topic_support an abstract transport method

1. all the transports need it (or there are crashes)
2. methods are easier to lint / check then properties
